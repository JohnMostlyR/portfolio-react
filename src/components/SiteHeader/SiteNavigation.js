import React, {Component} from 'react';
import {NavLink} from 'react-router-dom';
import {throttle} from 'lodash';
import 'scrolling-element';
import styled from 'styled-components';
import mq from '../../helpers/styles/templates/mediaQueries';
import typography from '../../helpers/styles/templates/typography';

const SiteNavigationNav = styled.nav`
  padding: 1vh 0;
  background-color: #575756;

  ${mq.m`
    left: 100%;
    position: absolute;
    top: 77px;
    transform: rotate(90deg);
    transform-origin: top left;
    background-color: transparent;
    border-bottom: none;  
  `}
`;

const SiteNavigationUL = styled.ul`
  display: flex;
  justify-content: space-around;
  list-style: none;
  margin: 0;
  padding: 0;

  ${mq.m`justify-content: flex-start;`}
`;

const SiteNavigationLI = styled.li`
  vertical-align: middle;

  ${mq.m`margin-top: 2vw;`}

  &:not(:first-child) {
    ${mq.m`margin-left: 1vh;`}
  }
`;

const SiteNavigationNavLink = styled(NavLink)`
  display: inline-block;
  position: relative;
  background-color: #FF6633;
  border-radius: 0;
  color: #fff;
  cursor: pointer;
  padding: 0.5em;
  text-align: center;
  text-decoration: none !important;
  transition: background-color 0.3s, color 0.3s;

  ${typography.pica}
  
  &:active,
  &:hover {
    background-color: transparent;
    color: #FF6633;

    &::before {
      opacity: 1;
      transform: scale3d(1, 1, 1);
    }
  }

  &::before {
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    top: 0;
    width: 100%;
    z-index: 1;
    border: 1px solid #FF6633;
    border-radius: inherit;
    opacity: 0;
    transform: scale3d(0.6, 0.6, 1);
    transition: transform 0.3s, opacity 0.3s;
    transition-timing-function: cubic-bezier(0.75, 0, 0.125, 1);

    ${mq.m`
      border: 2px solid #FF6633;
      border-radius: 5px;
    `}
  }

  ${mq.m`
    border-radius: 5px;
    width: 7rem; // For IE and Edge who don't support the max-content prop.
    width: max-content;
  `}
  
  &.is-active {
    background-color: transparent;
    color: #FF6633;

    &::before {
      opacity: 1;
      transform: scale3d(1, 1, 1);
    }
  }
`;

const SiteNavigationIcon = styled.span`
  display: none !important;

  ${mq.m`
    display: inherit !important;
    transform: rotate(-90deg);
    margin: 0;
  `}
`;

const SiteNavigationDescription = styled.span`
  ${mq.m`margin-left: .5rem;`}

  @media all and (orientation: landscape) and (min-width: 37.5em) {
    display: none;
  }

  @media all and (orientation: landscape) and (min-height: 37.5em) {
    display: inline;
  }
`;

class SiteNavigation extends Component {
  state = {
    myTopPosition: 0,
    topPositionOffset: 0,
  };

  _getElementTop = (element) => {
    let actualTop = element.offsetTop;
    let currentParent = element.offsetParent;

    while (currentParent !== null) {
      actualTop += currentParent.offsetTop;
      currentParent = currentParent.offsetParent;
    }

    return actualTop;
  };

  handleScrollEvent = (ev) => {
    let delta = (this.state.myTopPosition - ev.target.scrollingElement.scrollTop);
    if (delta <= 0) {
      this.setState({topPositionOffset: Math.abs(delta)});
    } else {
      this.setState({topPositionOffset: 0});
    }
  };

  componentDidMount() {
    const myTopPosition = this._getElementTop(this.siteNavigation);
    this.setState({myTopPosition});
    document.addEventListener('scroll', throttle(this.handleScrollEvent, 2));
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.handleScrollEvent);
  }

  render() {
    const pages = [
      {
        description: 'Intro',
        icon: 'fa fa-home fa-fw',
        navigateTo: 'intro',
      },
      {
        description: 'About',
        icon: 'fa fa-address-card fa-fw',
        navigateTo: 'about',
      },
      {
        description: 'Skills',
        icon: 'fa fa-code fa-fw',
        navigateTo: 'skills',
      },
      {
        description: 'Projects',
        icon: 'fa fa-list fa-fw',
        navigateTo: 'projects',
      },
      {
        description: 'Contact',
        icon: 'fa fa-send fa-fw',
        navigateTo: 'contact',
      },
    ];

    const setStyle = () => {
      if (this.state.topPositionOffset) {
        return (
          {
            position: 'absolute',
            top: (this.state.myTopPosition + this.state.topPositionOffset + 16),
            zIndex: 10,
          }
        );
      } else {
        return ({});
      }
    };

    return (
      <SiteNavigationNav style={setStyle()} innerRef={el => this.siteNavigation = el}>
        <SiteNavigationUL>
          {
            pages.map((page, idx) => {
              return (
                <SiteNavigationLI key={idx}>
                  <SiteNavigationNavLink
                    to={`/${page.navigateTo}`}
                    activeClassName='is-active'
                  ><SiteNavigationIcon
                    className={`${page.icon} ${this.props.className}`}
                    aria-hidden="true"/><span><SiteNavigationDescription>{page.description}</SiteNavigationDescription></span>
                  </SiteNavigationNavLink>
                </SiteNavigationLI>
              );
            })
          }
        </SiteNavigationUL>
      </SiteNavigationNav>
    );
  }
}

export default SiteNavigation;
