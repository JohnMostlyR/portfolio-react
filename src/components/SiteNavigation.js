import React, {Component} from 'react';
import {NavLink} from 'react-router-dom';
import {throttle} from 'lodash';
import 'scrolling-element';

class SiteNavigation extends Component {
  state = {
    myTopPosition: 0,
    topPositionOffset: 0,
  };

  _getElementTop = (element) => {
    let actualTop = element.offsetTop;
    let currentParent = element.offsetParent;

    while (currentParent !== null) {
      actualTop += currentParent.offsetTop;
      currentParent = currentParent.offsetParent;
    }

    return actualTop;
  };

  handleScrollEvent = (ev) => {
    let delta = (this.state.myTopPosition - ev.target.scrollingElement.scrollTop);
    if (delta <= 0) {
      this.setState({topPositionOffset: Math.abs(delta)});
    } else {
      this.setState({topPositionOffset: 0});
    }
  };
  componentDidMount() {
    const myTopPosition = this._getElementTop(this.refs.siteNavigation);
    this.setState({myTopPosition});
    document.addEventListener('scroll', throttle(this.handleScrollEvent, 2));
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.handleScrollEvent);
  }

  render() {
    const pages = [
      {
        description: 'Intro',
        icon: 'fa fa-home fa-fw',
        navigateTo: 'intro',
      },
      {
        description: 'About',
        icon: 'fa fa-address-card fa-fw',
        navigateTo: 'about',
      },
      {
        description: 'Skills',
        icon: 'fa fa-code fa-fw',
        navigateTo: 'skills',
      },
      {
        description: 'Projects',
        icon: 'fa fa-list fa-fw',
        navigateTo: 'projects',
      },
      {
        description: 'Contact',
        icon: 'fa fa-send fa-fw',
        navigateTo: 'contact',
      },
    ];

    const setStyle = () => {
      if (this.state.topPositionOffset) {
        return (
          {
            position: 'absolute',
            top: (this.state.myTopPosition + this.state.topPositionOffset),
            width: '100%',
            zIndex: 10,
          }
        );
      } else {
        return ({});
      }
    };

    return (
      <nav style={setStyle()} className="c-portfolio-site-nav" ref='siteNavigation'>
        <ul className="c-portfolio-site-nav__items">
          {
            pages.map((page, idx) => {
              return (
                <li
                  className="c-portfolio-site-nav__item"
                  key={idx}
                >
                  <NavLink
                    to={`/${page.navigateTo}`}
                    activeClassName='is-active'
                    className='c-portfolio-site-nav__link'
                  ><span
                    className={`${page.icon} c-portfolio-site-nav__icon`}
                    aria-hidden="true"/><span><span
                    className="c-portfolio-site-nav__description">{page.description}</span></span>
                  </NavLink>
                </li>
              );
            })
          }
        </ul>
      </nav>
    );
  }
}

export default SiteNavigation;
