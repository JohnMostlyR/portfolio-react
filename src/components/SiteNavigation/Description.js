import styled from 'styled-components';
import mq from '../../styles/templates/mediaQueries';

const Description = styled.span`
  ${mq.m`margin-left: .5rem;`}

  @media all and (orientation: landscape) and (min-width: 37.5em) {
    display: none;
  }

  @media all and (orientation: landscape) and (min-height: 37.5em) {
    display: inline;
  }
`;

export default Description;