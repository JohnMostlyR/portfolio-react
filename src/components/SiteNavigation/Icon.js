import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import mq from '../../styles/templates/mediaQueries';

const SiteNavigationIcon = styled.span`
  display: none !important;

  ${mq.m`
    display: inherit !important;
    transform: rotate(-90deg);
    margin: 0;
  `}
`;

const Icon = (props) => (
  <SiteNavigationIcon className={props.iconClassName} aria-hidden="true"><span>{props.children}</span></SiteNavigationIcon>
);

Icon.propTypes = {
  iconClassName: PropTypes.string,
};

export default Icon;