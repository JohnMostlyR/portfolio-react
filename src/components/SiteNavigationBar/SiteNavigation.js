import styled from 'styled-components';
import {rem} from 'polished';

import mq from '../../styles/templates/mediaQueries';

const SiteNavigation = styled.nav`
  width: 100%;
  padding: .35rem 0;
  position: ${props => props.isAtScreenTop ? 'fixed' : 'unset'};
  top: 0;
  background-color: #575756;

  ${mq.m`
    left: 100%;
    position: ${props => props.isAtScreenTop ? 'fixed' : 'absolute'};
    top: ${props => props.isAtScreenTop ? 0 : rem('77px')};
    width: unset;
    padding: 0 .35rem;
    transform: rotate(90deg);
    transform-origin: top left;
    background-color: transparent;
    border-bottom: none;  
  `}
`;

export default SiteNavigation;
