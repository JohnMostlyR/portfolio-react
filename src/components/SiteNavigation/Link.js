import React from 'react';
import PropTypes from 'prop-types';
import {NavLink} from 'react-router-dom';
import styled from 'styled-components';
import mq from '../../styles/templates/mediaQueries';
import typography from '../../styles/templates/typography';

import Icon from './Icon';
import Description from './Description';

const SiteNavLink = styled(NavLink)`
  display: inline-block;
  position: relative;
  background-color: #FF6633;
  border-radius: 0;
  color: #fff;
  cursor: pointer;
  padding: 0.5em;
  text-align: center;
  text-decoration: none !important;
  transition: background-color 0.3s, color 0.3s;

  ${typography.pica}
  
  &:active,
  &:hover {
    background-color: transparent;
    color: #FF6633;

    &::before {
      opacity: 1;
      transform: scale3d(1, 1, 1);
    }
  }

  &::before {
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    top: 0;
    width: 100%;
    z-index: 1;
    border: 1px solid #FF6633;
    border-radius: inherit;
    opacity: 0;
    transform: scale3d(0.6, 0.6, 1);
    transition: transform 0.3s, opacity 0.3s;
    transition-timing-function: cubic-bezier(0.75, 0, 0.125, 1);

    ${mq.m`
      border: 2px solid #FF6633;
      border-radius: 5px;
    `}
  }

  ${mq.m`
    border-radius: 5px;
    width: 7rem; // For IE and Edge who don't support the max-content prop.
    width: max-content;
  `}
  
  &.is-active {
    background-color: transparent;
    color: #FF6633;

    &::before {
      opacity: 1;
      transform: scale3d(1, 1, 1);
    }
  }
`;

const Link = (props) => (
  <SiteNavLink to={props.to} activeClassName="is-active">
    <Icon iconClassName={props.iconClassName}/><span><Description>{props.description}</Description></span>
  </SiteNavLink>
);

Link.propTypes = {
  to: PropTypes.string,
  iconClassName: PropTypes.string,
  description: PropTypes.string,
};

Link.defaultProps = {
  to: '#',
};

export default Link;