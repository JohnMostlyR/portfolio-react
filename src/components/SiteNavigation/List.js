import React from 'react';
import styled from 'styled-components';
import mq from '../../styles/templates/mediaQueries';

import ListItem from './ListItem';
import Link from './Link';

const UL = styled.ul`
  display: flex;
  justify-content: space-around;
  list-style: none;
  margin: 0;
  padding: 0;

  ${mq.m`justify-content: flex-start;`}
`;

const pages = [
  {
    description: 'Home',
    icon: 'fa fa-home fa-fw',
    navigateTo: 'home',
  },
  {
    description: 'About',
    icon: 'fa fa-address-card fa-fw',
    navigateTo: 'about',
  },
  {
    description: 'Skills',
    icon: 'fa fa-code fa-fw',
    navigateTo: 'skills',
  },
  {
    description: 'Projects',
    icon: 'fa fa-list fa-fw',
    navigateTo: 'projects',
  },
  {
    description: 'Contact',
    icon: 'fa fa-send fa-fw',
    navigateTo: 'contact',
  },
];

const List = (props) => (
  <UL>
    {
      pages.map((page, idx) => (
          <ListItem key={idx}>
            <Link to={`/${page.navigateTo}`} iconClassName={page.icon}
                  description={page.description}/>
          </ListItem>
        ),
      )
    }
  </UL>
);

export default List;

