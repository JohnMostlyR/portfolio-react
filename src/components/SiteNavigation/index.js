import React, {Component} from 'react';
import {throttle} from 'lodash';
// import 'scrolling-element';
import styled from 'styled-components';
import mq from '../../styles/templates/mediaQueries';

import List from './List';

const SiteNavigationNav = styled.nav`
  padding: 1vh 0;
  background-color: #575756;

  ${mq.m`
    left: 100%;
    position: absolute;
    top: 77px;
    transform: rotate(90deg);
    transform-origin: top left;
    background-color: transparent;
    border-bottom: none;  
  `}
`;

class SiteNavigation extends Component {
  state = {
    myTopPosition: 0,
    topPositionOffset: 0,
  };

  _getElementTop = (element) => {
    let actualTop = element.offsetTop;
    let currentParent = element.offsetParent;

    while (currentParent !== null) {
      actualTop += currentParent.offsetTop;
      currentParent = currentParent.offsetParent;
    }

    return actualTop;
  };

  handleScrollEvent = (ev) => {
    let delta = (this.state.myTopPosition - ev.target.scrollingElement.scrollTop);
    if (delta <= 0) {
      this.setState({topPositionOffset: Math.abs(delta)});
    } else {
      this.setState({topPositionOffset: 0});
    }
  };

  componentDidMount() {
    const myTopPosition = this._getElementTop(this.siteNavigation);
    this.setState({myTopPosition});
    document.addEventListener('scroll', throttle(this.handleScrollEvent, 2));
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.handleScrollEvent);
  }

  render() {
    const setStyle = () => {
      if (this.state.topPositionOffset) {
        return (
          {
            position: 'absolute',
            top: (this.state.myTopPosition + this.state.topPositionOffset + 16),
            zIndex: 10,
          }
        );
      } else {
        return ({});
      }
    };

    return (
      <SiteNavigationNav style={setStyle()} innerRef={el => this.siteNavigation = el}>
        <List/>
      </SiteNavigationNav>
    );
  }
}

export default SiteNavigation;
