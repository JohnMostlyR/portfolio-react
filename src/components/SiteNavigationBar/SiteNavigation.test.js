import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import SiteNavigation from './SiteNavigation';

describe('SiteNavigation', () => {
  it('Should render and match the snapshot', () => {
    const tree = renderer.create(<SiteNavigation/>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});