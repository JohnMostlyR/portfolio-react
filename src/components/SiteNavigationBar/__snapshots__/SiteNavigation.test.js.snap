// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiteNavigation Should render and match the snapshot 1`] = `
.c0 {
  width: 100%;
  padding: .35rem 0;
  position: unset;
  top: 0;
  background-color: #575756;
}

@media (min-width:37.5em) {
  .c0 {
    left: 100%;
    position: absolute;
    top: 4.8125rem;
    width: unset;
    padding: 0 .35rem;
    -webkit-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
    -webkit-transform-origin: top left;
    -ms-transform-origin: top left;
    transform-origin: top left;
    background-color: transparent;
    border-bottom: none;
  }
}

<nav
  className="c0"
/>
`;
